feat: introduce new abandoned shopping cart topic

Ami-Thomas opened the pull request:
This pull request add the generation of events for a new topic:

SHOPPING.CARTS.ABANDONED: events for abandoned shopping carts


Tested locally and updated with new message format

<img width="1611" alt="Screenshot 2025-04-24 at 8 03 29 PM" src="https://github.com/user-attachments/assets/17b42e04-b28a-4fba-87f2-8d790a23ecfd" />

`{
  "cartid": "c5a4e2fc-13ee-4fde-b4f2-4a22b565d09d",
  "customer": {
    "id": "573b6b1d-6070-4f4b-afb1-be480554930a",
    "name": "Inger Orn",
    "emails": [
      "inger.orn@example.com"
    ]
  },
  "products": [
    "Navy Capri Jeans",
    "Black Cargo Jeans"
  ],
  "abandonedtime": "2025-04-24T20:01:32.136Z"
}`

noelthekk commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2024 IBM Corp. All Rights Reserved.
```
Please update it to 2025

Ami-Thomas replied:
fixed

noelthekk commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/generators/AbandonedShoppingCartGenerator.java:
```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2024 IBM Corp. All Rights Reserved.
```
Same as above.

Ami-Thomas replied:
fixed

noelthekk reviewed (commented):
Few minor comments

noelthekk commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/tasks/AbandonedShoppingCartsTask.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright 2023 IBM Corp. All Rights Reserved.
```
Same as above.

Ami-Thomas replied:
fixed

rachel-orti commented on README.md:
```diff
@@ -4,18 +4,19 @@ Kafka Connect source connector used for generating simulated events for demos an
+| `STOCK.NOSTOCK`            | A product has run out-of-stock                                                                |
+| `PRODUCT.RETURNS`          | A return request has been issued                                                              |
+| `PRODUCT.REVIEWS`          | A product review has been posted                                                              |
+| `ABANDONED.SHOPPING.CARTS` | A customer has added items to their shopping cart but left it without completing the purchase |
```
Usually we are putting the main subject first and then the adjective. For example ORDERS.ONLINE, CUSTOMERS.NEW, etc. So here it should be SHOPPING.CARTS.ABANDONED to be consistent.

As it is "a customer" (that is singular), it should be "his/her" shopping cart (instead of "their").

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+    private final String cartId;
+
+    /** Time that the cart was considered abandoned */
+    private final String abandonTime;
```
To be consistent with other classes, we can simply call it `timestamp` (see `OnlineOrder` for example). The current class is called `AbandonedShoppingCart` so the timestamp is the abandoned timestamp as there is only one timestamp in the class.
If you want to use something more specific, it should be: `abandonedTimestamp` or `abandonmentTimestamp` as abandon is a verb and not a noun.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+    private static final Schema SCHEMA = SchemaBuilder.struct()
+            .name("abandonedshoppingcart")
+            .version(1)
+            .field("cartId",      Schema.STRING_SCHEMA)
```
To be consistent with other events, we should not use uppercases for the schema fields. So it should be:
- cartid
- abandonedtime (or abandonmenttime)

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+            .build();
+
+    /** Creates an {@link AbandonedShoppingCart} using the provided details */
+    public AbandonedShoppingCart(String cartId, String abandonTime, OnlineCustomer customer, List<Product> products, ZonedDateTime recordTimestamp) {
```
Replace "abandonTime" to use the same name that is used for the property in the class depending on what you have chosen to do. See previous comment. Could be simply "timestamp" or "abandonedTimestamp" or "abandonmentTimestamp".

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+        this.products = products;
+    }
+
+    /** Creates an {@link OnlineOrder} using the provided details.
```
It should be AbandonedShoppingCart instead of OnlineOrder.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+    /** Creates an {@link OnlineOrder} using the provided details.
+     * The ID is generated randomly.
+     * */
+    public AbandonedShoppingCart(String abandonTime, OnlineCustomer customer, List<Product> products, ZonedDateTime recordTimestamp) {
```
Replace "abandonTime" to use the same name that is used for the property in the class depending on what you have chosen to do. See previous comment. Could be simply "timestamp" or "abandonedTimestamp" or "abandonmentTimestamp".

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+        this(UUID.randomUUID().toString(), abandonTime, customer, products, recordTimestamp);
+    }
+
+    public String getAbandonTime() {
```
Replace "abandonTime" to use the same name that is used for the property in the class depending on what you have chosen to do. See previous comment. Could be simply "timestamp" or "abandonedTimestamp" or "abandonmentTimestamp".

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+        return cartId;
+    }
+
+    public String getTimestamp() {
```
If you decide to use "abondonedTimestamp" or "abandonmentTimestamp" instead of "timestamp" you should remove this method.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+        struct.put(SCHEMA.field("cartId"),      cartId);
+        struct.put(SCHEMA.field("customer"),    customer.toStruct());
+        struct.put(SCHEMA.field("products"),    products.stream().map(Product::toStruct).collect(Collectors.toList()));
+        struct.put(SCHEMA.field("abandonTime"), abandonTime);
```
Once the update will be done for the schema (see previous comment), update this code that sets the abandoned time accordingly.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/data/AbandonedShoppingCart.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Override
+    public String toString() {
+        return "AbandonedShoppingCart [cartId=" + cartId + ", abandonTime=" + abandonTime + ", customer=" + customer + ", products="
```
Update the property name for the abandoned timestamp (see previous messages).

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -259,6 +269,13 @@ public class DatagenSourceConfig {
                     CONFIG_GROUP_TOPICNAMES, 10, Width.LONG, "Product reviews topic")
+        .define(CONFIG_TOPICNAME_ABANDONEDSHOPPINGCARTS,
+                    Type.STRING,
+                    "ABANDONED.SHOPPING.CARTS",
```
Here it should be SHOPPING.CARTS.ABANDONED to be consistent with the other topic names.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -259,6 +269,13 @@ public class DatagenSourceConfig {
+                    new NonEmptyString(),
+                    Importance.LOW,
+                    "Name of the topic to use for abandoned shopping cart events",
+                    CONFIG_GROUP_TOPICNAMES, 10, Width.LONG, "Abandoned shopping carts topic")
```
It should be 11 instead of 10. This value must be incremented each time we add a new topic.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -756,6 +804,13 @@ public class DatagenSourceConfig {
+                    Range.between(0, 900),  // up to 15 mins max
+                    Importance.LOW,
+                    "Maximum delay (in *seconds*) to produce new abandoned shopping cart events (this is the maximum difference allowed between the timestamp string in the event payload, and the Kafka message's metadata timestamp)",
+                    CONFIG_GROUP_DELAYS, 10, Width.SHORT, "Abandoned shopping cart events - max produce delay")
```
It should be 11 instead of 10. This value must be incremented each time we add a new topic.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -829,6 +884,13 @@ public class DatagenSourceConfig {
+                    Range.between(0.0, 1.0), // ratio should be between 0 (don't create duplicates) and 1 (duplicate every message)
+                    Importance.LOW,
+                    "Ratio of abandoned shopping cart events that should be duplicated. Must be between 0 and 1.",
+                    CONFIG_GROUP_DUPLICATES, 10, Width.SHORT, "Duplicate abandoned shopping cart events ratio")
```
It should be 11 instead of 10. This value must be incremented each time we add a new topic.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -909,6 +971,13 @@ public class DatagenSourceConfig {
+                    Range.atLeast(500),
+                    Importance.LOW,
+                    "Delay, in milliseconds, between each abandoned shopping cart that should be generated.",
+                    CONFIG_GROUP_TIMES, 11, Width.MEDIUM, "Abandoned carts delay")
```
It should be 12 instead of 11. This value must be incremented each time we add a new topic.

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -909,6 +971,13 @@ public class DatagenSourceConfig {
                     CONFIG_GROUP_TIMES, 11, Width.MEDIUM, "Product reviews delay")
+        .define(CONFIG_TIMES_ABANDONEDSHOPPINGCARTS,
+                    Type.INT,
+                    30_000, // 30 seconds
```
I have noticed that - by default - we have exactly the same rate for "online orders" and for "abandoned shopping rate". Is it the expected behaviour ?

Ami-Thomas replied:
Like orders and online orders, I added 30 secs value. But in real use case, users don’t abandon carts every 30 seconds, it could be >15mins. But for demoing purpose, I changed it to 5 mins

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/generators/AbandonedShoppingCartGenerator.java:
```diff
@@ -0,0 +1,79 @@
+    protected AbandonedShoppingCart generateEvent(ZonedDateTime timestamp) {
+        // Generate a random customer
+        OnlineCustomer customer = OnlineCustomer.create(faker, minEmails, maxEmails);
+        String abandonTime = formatTimestamp(timestamp);
```
The abandonTime variable is not necessary here and it is far from where it is used. It would be better to directly call `formatTimestamp(timestamp)` when calling the constructor line 77.

rachel-orti commented on README.md:
```diff
@@ -368,6 +369,17 @@ spec:
     productreviews.delay.ms.min: 300000   # 5 minutes
     productreviews.delay.ms.max: 3600000  # 1 hour
 
+    #
```
Some settings for abandoned shopping carts are missing in the README: topic name, timings, event delays, duplicates,

rachel-orti commented on README.md:
```diff
@@ -368,6 +369,17 @@ spec:
+    #
+    # abandoned shopping carts
+    #
+    # these events are intended to represent shopping carts that were abandoned before checkout,
```
It should be a dot instead of a comma after the checkout word.

rachel-orti commented on README.md:
```diff
@@ -368,6 +369,17 @@ spec:
+    abandonedshoppingcarts.products.max: 5
+    # number of emails for the customer who abandoned the shopping cart: between 1 and 2 (inclusive)
+    abandonedshoppingcarts.customer.emails.min: 1
+    abandonedshoppingcarts.customer.emails.max: 2
```
It would be better to add an empty line here (before the regions definition).

rachel-orti commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenHistoryGenerator.java:
```diff
@@ -153,26 +155,16 @@ private void addStockMovementRecords(List<SourceRecord> historicalRecords, Abstr
-        final String OUTOFSTOCK_TOPIC = config.getString(DatagenSourceConfig.CONFIG_TOPICNAME_OUTOFSTOCKS);
+    private void addAbandonedShoppingCartRecords(List<SourceRecord> historicalRecords, AbstractConfig config) {
+        log.debug("generating historical abandoned shopping cart records");
+        final String TOPIC = config.getString(DatagenSourceConfig.CONFIG_TOPICNAME_ABANDONEDSHOPPINGCARTS);
```
The code for `addAbandonedShoppingCartRecords` is wrong. Here we should not create OnlineOrder events:
![image](https://github.com/user-attachments/assets/80611bb6-d2d5-49fa-a47e-ca2db7707c4c)

rachel-orti reviewed (changes requested):
I have added some comments.

rachel-orti reviewed (commented):
You should also add some documentation for the new topic in the doc folder see section "Step 7: Document" in the following box note: https://ibm.ent.box.com/notes/1492922600913?s=cym13i28n5vsdiasoe3mo01w2md5adlz

rachel-orti commented on README.md:
```diff
@@ -183,25 +189,27 @@ spec:
     # product reviews
-    duplicates.productreviews.ratio: 0      # events not duplicated
+    duplicates.productreviews.ratio: 0          # events not duplicated
+    # online orders
```
Here it should be "abandoned shopping carts" in the comment.

Ami-Thomas replied:
changed

rachel-orti reviewed (changes requested):
I have added one comment.

rachel-orti reviewed (commented):
It is OK now. I will approve when you will have removed the JAR that was committed for testing purpose as we should not merge it.

rachel-orti reviewed (approved):
It is OK now.

dalelane commented on README.md:
```diff
@@ -4,18 +4,19 @@ Kafka Connect source connector used for generating simulated events for demos an
+| `STOCK.NOSTOCK`            | A product has run out-of-stock                                                                  |
+| `PRODUCT.RETURNS`          | A return request has been issued                                                                |
+| `PRODUCT.REVIEWS`          | A product review has been posted                                                                |
+| `SHOPPING.CARTS.ABANDONED` | A customer has added items to his/her shopping cart but left it without completing the purchase |
```
```suggestion
| `ORDERS.ABANDONED` | A customer has added items to his/her shopping cart but left it without completing the purchase |
```

Naming is naturally subjective, so there are no definitive answers here, in this context a three-layer name feels a little excessive in my opinion. 

Keeping in mind that the convention with topic names is similar to something like a Java package, I think extending the existing `ORDERS.*` topics would be a cleaner fit.

Ami-Thomas replied:
changed to ORDERS.ABANDONED

dalelane commented on README.md:
```diff
@@ -82,6 +83,7 @@ spec:
     topic.name.outofstocks: STOCK.NOSTOCK
     topic.name.returnrequests: PRODUCT.RETURNS
     topic.name.productreviews: PRODUCT.REVIEWS
+    topic.name.abandonedshoppingcarts: SHOPPING.CARTS.ABANDONED
```
```suggestion
    topic.name.abandonedshoppingcarts: ORDERS.ABANDONED
```

as above

Ami-Thomas replied:
changed. Renamed abandoned shopping cart occurrences to abandoned orders

dalelane commented on doc/SHOPPING.CARTS.ABANDONED/schema.avro:
```diff
@@ -0,0 +1,59 @@
+    },
+    {
+      "name": "abandonedtime",
+      "doc": "Time that the products were abandoned in the shopping cart (UTC time in ISO 8601 format)",
```
What does "abandoned" mean in this context? How can a basket be abandoned at a specific time? 

e.g. if I don't add anything to my basket for 5 minutes, is that abandoned? I could come back and add something after 6 minutes. 

I think the simplest update (to avoid needing to make corresponding code changes) could be to update this to clarify that this is the time that the basket was determined to be abandoned after some period (e.g. an hour?) of inactivity.

Ami-Thomas replied:
changed wording to "Time at which the shopping cart was determined to be abandoned after a period of inactivity (UTC time in ISO 8601 format)."

dalelane commented on pom.xml:
```diff
@@ -8,7 +8,7 @@
     <artifactId>kafka-connect-loosehangerjeans-source</artifactId>
     <description>Kafka Connect source connector used for generating simulated events for demos and tests</description>
-    <version>0.2.1</version>
+    <version>0.2.2</version>
```
```suggestion
    <version>0.3.0</version>
```

As you're adding a new topic, I think a minor version number change would be more appropriate.

Ami-Thomas replied:
ok, done

dalelane commented on doc/SHOPPING.CARTS.ABANDONED/schema.avro:
```diff
@@ -0,0 +1,59 @@
+          {
+            "name": "emails",
+            "type": {
+              "type": "array",
```
~Why multiple email addresses? Do we need this for the scenario?~

~Multiple different email addresses feels like an odd thing to associate with a shopping basket, so I'm curious about the thinking behind this~

ah.... I've read a bit further into the PR now, so I understand that this is a by-product of re-using the schema from online orders - so never mind! :)

dalelane commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -259,6 +269,13 @@ public class DatagenSourceConfig {
                     CONFIG_GROUP_TOPICNAMES, 10, Width.LONG, "Product reviews topic")
+        .define(CONFIG_TOPICNAME_ABANDONEDSHOPPINGCARTS,
+                    Type.STRING,
+                    "SHOPPING.CARTS.ABANDONED",
```
```suggestion
                    "ORDERS.ABANDONED",
```

as above

Ami-Thomas replied:
fixed

dalelane commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConfig.java:
```diff
@@ -684,6 +701,37 @@ public class DatagenSourceConfig {
+                    1,
+                    Range.atLeast(1),
+                    Importance.LOW,
+                    "Minimum number of products in an abandoned shopping cart. Must be greater than 0.",
```
I'm curious about the reason for "must be greater than 0" - would allowing for empty baskets cause problems?

Ami-Thomas replied:
My thought was to produce atleast one abandoned product in an order (as the topic name suggests)

dalelane commented on src/main/java/com/ibm/eventautomation/demos/loosehangerjeans/DatagenSourceConnector.java:
```diff
@@ -27,7 +27,7 @@
 public class DatagenSourceConnector extends SourceConnector {
 
-    protected static final String VERSION = "0.2.1";
+    protected static final String VERSION = "0.2.2";
```
```suggestion
    protected static final String VERSION = "0.3.0";
```

as above

Ami-Thomas replied:
fixed

dalelane reviewed (changes requested):
I have a couple questions and minor suggestions, but this is looking good

dalelane commented:
thanks for doing the rebase 👍

Commit:
feat: add new abandoned orders topic

add new abandoned orders topic

Contributes to: https://jsw.ibm.com/browse/EVI-11776

Signed-off-by: Ami Thomas <Ami.Thomas@ibm.com>

Commit:
feat: address formatting changes

address formatting changes

Contributes to: https://jsw.ibm.com/browse/EVI-11776

Signed-off-by: Ami Thomas <Ami.Thomas@ibm.com>
